### ISSUE TITLE
Forwarder getting stuck in bad state when queue gets too big

### ISSUE THREAD

**Issue by conorbranagan** [DataDog/dd-agent#943]
Not 100% clear how it gets into this state in the first place, but we don't seem to ever recover when the queue gets too large. Snippet from the logs:

```
2014-05-10 03:58:35 UTC | DEBUG | dd.forwarder | util(util.py:354) | Resetting watchdog for 50000
2014-05-10 03:58:35 UTC | DEBUG | dd.forwarder | transaction(transaction.py:136) | A flush is already in progress, not doing anything
2014-05-10 03:58:36 UTC | DEBUG | dd.forwarder | transaction(transaction.py:112) | New transaction to add, total size of queue would be: 30720 KB
2014-05-10 03:58:36 UTC | WARNING | dd.forwarder | transaction(transaction.py:115) | Queue is too big, removing old transactions...
2014-05-10 03:58:36 UTC | WARNING | dd.forwarder | transaction(transaction.py:122) | Removed transaction 2685 from queue
2014-05-10 03:58:36 UTC | DEBUG | dd.forwarder | transaction(transaction.py:130) | Transaction 2686 added
2014-05-10 03:58:36 UTC | DEBUG | dd.forwarder | transaction(transaction.py:97) | Queue size: at 1399694316.49, 703 transaction(s), 30719 KB
2014-05-10 03:58:36 UTC | DEBUG | dd.forwarder | forwarder(dd-forwarder:172) | Created transaction 2686
2014-05-10 03:58:36 UTC | DEBUG | dd.forwarder | transaction(transaction.py:136) | A flush is already in progress, not doing anything
2014-05-10 03:58:36 UTC | DEBUG | dd.forwarder | forwarder(dd-forwarder:339) | 200 POST /intake?api_key=REMOVED (127.0.0.1) 4.81ms
2014-05-10 03:58:40 UTC | DEBUG | dd.forwarder | util(util.py:354) | Resetting watchdog for 50000
2014-05-10 03:58:40 UTC | DEBUG | dd.forwarder | transaction(transaction.py:136) | A flush is already in progress, not doing anything
2014-05-10 03:58:41 UTC | DEBUG | dd.forwarder | transaction(transaction.py:112) | New transaction to add, total size of queue would be: 30720 KB
2014-05-10 03:58:41 UTC | WARNING | dd.forwarder | transaction(transaction.py:115) | Queue is too big, removing old transactions...
2014-05-10 03:58:41 UTC | WARNING | dd.forwarder | transaction(transaction.py:122) | Removed transaction 2686 from queue
2014-05-10 03:58:41 UTC | DEBUG | dd.forwarder | transaction(transaction.py:130) | Transaction 2687 added
2014-05-10 03:58:41 UTC | DEBUG | dd.forwarder | transaction(transaction.py:97) | Queue size: at 1399694321.86, 703 transaction(s), 30714 KB
2014-05-10 03:58:41 UTC | DEBUG | dd.forwarder | forwarder(dd-forwarder:172) | Created transaction 2687
2014-05-10 03:58:41 UTC | DEBUG | dd.forwarder | transaction(transaction.py:136) | A flush is already in progress, not doing anything
2014-05-10 03:58:41 UTC | DEBUG | dd.forwarder | forwarder(dd-forwarder:339) | 200 POST /api/v1/series?api_key=REMOVED (127.0.0.1) 3.09ms
2014-05-10 03:58:45 UTC | DEBUG | dd.forwarder | util(util.py:354) | Resetting watchdog for 50000
2014-05-10 03:58:45 UTC | DEBUG | dd.forwarder | transaction(transaction.py:136) | A flush is already in progress, not doing anything
2014-05-10 03:58:50 UTC | DEBUG | dd.forwarder | util(util.py:354) | Resetting watchdog for 50000
2014-05-10 03:58:50 UTC | DEBUG | dd.forwarder | transaction(transaction.py:136) | A flush is already in progress, not doing anything
2014-05-10 03:58:51 UTC | DEBUG | dd.forwarder | transaction(transaction.py:112) | New transaction to add, total size of queue would be: 30715 KB
2014-05-10 03:58:51 UTC | DEBUG | dd.forwarder | transaction(transaction.py:130) | Transaction 2688 added
2014-05-10 03:58:51 UTC | DEBUG | dd.forwarder | transaction(transaction.py:97) | Queue size: at 1399694331.92, 704 transaction(s), 30715 KB
2014-05-10 03:58:51 UTC | DEBUG | dd.forwarder | forwarder(dd-forwarder:172) | Created transaction 2688
2014-05-10 03:58:51 UTC | DEBUG | dd.forwarder | transaction(transaction.py:136) | A flush is already in progress, not doing anything
2014-05-10 03:58:51 UTC | DEBUG | dd.forwarder | forwarder(dd-forwarder:339) | 200 POST /api/v1/series?api_key=REMOVED (127.0.0.1) 2.24ms
2014-05-10 03:58:55 UTC | DEBUG | dd.forwarder | util(util.py:354) | Resetting watchdog for 50000
2014-05-10 03:58:55 UTC | DEBUG | dd.forwarder | transaction(transaction.py:136) | A flush is already in progress, not doing anything
2014-05-10 03:58:56 UTC | DEBUG | dd.forwarder | transaction(transaction.py:112) | New transaction to add, total size of queue would be: 30720 KB
2014-05-10 03:58:56 UTC | WARNING | dd.forwarder | transaction(transaction.py:115) | Queue is too big, removing old transactions...
2014-05-10 03:58:56 UTC | WARNING | dd.forwarder | transaction(transaction.py:122) | Removed transaction 2688 from queue
2014-05-10 03:58:56 UTC | WARNING | dd.forwarder | transaction(transaction.py:122) | Removed transaction 2687 from queue
2014-05-10 03:58:56 UTC | DEBUG | dd.forwarder | transaction(transaction.py:130) | Transaction 2689 added
2014-05-10 03:58:56 UTC | DEBUG | dd.forwarder | transaction(transaction.py:97) | Queue size: at 1399694336.47, 703 transaction(s), 30719 KB
2014-05-10 03:58:56 UTC | DEBUG | dd.forwarder | forwarder(dd-forwarder:172) | Created transaction 2689
2014-05-10 03:58:56 UTC | DEBUG | dd.forwarder | transaction(transaction.py:136) | A flush is already in progress, not doing anything
2014-05-10 03:58:56 UTC | DEBUG | dd.forwarder | forwarder(dd-forwarder:339) | 200 POST /intake?api_key=REMOVED (127.0.0.1) 4.73ms
2014-05-10 03:59:00 UTC | DEBUG | dd.forwarder | util(util.py:354) | Resetting watchdog for 50000
2014-05-10 03:59:00 UTC | DEBUG | dd.forwarder | transaction(transaction.py:136) | A flush is already in progress, not doing anything
2014-05-10 03:59:01 UTC | DEBUG | dd.forwarder | transaction(transaction.py:112) | New transaction to add, total size of queue would be: 30720 KB
2014-05-10 03:59:01 UTC | WARNING | dd.forwarder | transaction(transaction.py:115) | Queue is too big, removing old transactions...
2014-05-10 03:59:01 UTC | WARNING | dd.forwarder | transaction(transaction.py:122) | Removed transaction 2689 from queue
2014-05-10 03:59:01 UTC | DEBUG | dd.forwarder | transaction(transaction.py:130) | Transaction 2690 added
2014-05-10 03:59:01 UTC | DEBUG | dd.forwarder | transaction(transaction.py:97) | Queue size: at 1399694341.99, 703 transaction(s), 30714 KB
2014-05-10 03:59:01 UTC | DEBUG | dd.forwarder | forwarder(dd-forwarder:172) | Created transaction 2690
2014-05-10 03:59:01 UTC | DEBUG | dd.forwarder | transaction(transaction.py:136) | A flush is already in progress, not doing anything
2014-05-10 03:59:01 UTC | DEBUG | dd.forwarder | forwarder(dd-forwarder:339) | 200 POST /api/v1/series?api_key=REMOVED(127.0.0.1) 7.92ms
```

**Comment by remh** [comment]
Closing this issue. I was unable reproduce the issue. 
I was able to recover from a 800 payloads queue without a blip.

**Comment by mccutchen** [comment]
We just ran into what looks like the same behavior on a production host, which stopped reporting metrics 4 days ago.  Unfortunately, it took us until today to notice the lack of metrics from this one host.

Also unfortunately, we do not have debug logs turned on for the agent.  The logs are filled with entries that look like this:

```
$ tail -15 /var/log/datadog/forwarder.log
2017-03-27 18:02:31 UTC | WARNING | dd.forwarder | transaction(transaction.py:133) | Queue is too big, removing old transactions...
2017-03-27 18:02:31 UTC | WARNING | dd.forwarder | transaction(transaction.py:140) | Removed transaction 90225 from queue
2017-03-27 18:02:31 UTC | WARNING | dd.forwarder | transaction(transaction.py:140) | Removed transaction 90224 from queue
2017-03-27 18:02:31 UTC | WARNING | dd.forwarder | transaction(transaction.py:140) | Removed transaction 90223 from queue
2017-03-27 18:02:50 UTC | WARNING | dd.forwarder | transaction(transaction.py:133) | Queue is too big, removing old transactions...
2017-03-27 18:02:50 UTC | WARNING | dd.forwarder | transaction(transaction.py:140) | Removed transaction 90228 from queue
2017-03-27 18:02:50 UTC | WARNING | dd.forwarder | transaction(transaction.py:140) | Removed transaction 90227 from queue
2017-03-27 18:02:50 UTC | WARNING | dd.forwarder | transaction(transaction.py:140) | Removed transaction 90226 from queue
2017-03-27 18:03:10 UTC | WARNING | dd.forwarder | transaction(transaction.py:133) | Queue is too big, removing old transactions...
2017-03-27 18:03:10 UTC | WARNING | dd.forwarder | transaction(transaction.py:140) | Removed transaction 90231 from queue
2017-03-27 18:03:10 UTC | WARNING | dd.forwarder | transaction(transaction.py:140) | Removed transaction 90230 from queue
2017-03-27 18:03:10 UTC | WARNING | dd.forwarder | transaction(transaction.py:140) | Removed transaction 90229 from queue
2017-03-27 18:03:29 UTC | WARNING | dd.forwarder | transaction(transaction.py:133) | Queue is too big, removing old transactions...
2017-03-27 18:03:29 UTC | WARNING | dd.forwarder | transaction(transaction.py:140) | Removed transaction 90233 from queue
2017-03-27 18:03:29 UTC | WARNING | dd.forwarder | transaction(transaction.py:140) | Removed transaction 90232 from queue
```

Running `ddagent.py info` on the affected host yields this output:

```
$ sudo /opt/datadog-agent/agent/ddagent.py info
====================
Forwarder (v 5.11.2)
====================

  Status date: 2017-03-23 12:10:08 (20343s ago)
  Pid: 2309
  Platform: Linux-4.2.0-41-generic-x86_64-with-Ubuntu-14.04-trusty
  Python Version: 2.7.12, 64bit
  Logs: <stderr>, /var/log/datadog/forwarder.log, syslog:/dev/log

  Queue Size: 31124285 bytes
  Queue Length: 188
  Flush Count: 75615
  Transactions received: 34501
  Transactions flushed: 34312
  Transactions rejected: 0
  API Key Status: API Key is valid
```

A couple of things to note in that `info` output:
- the status date is (as of this writing) more than 5 hours old
- none of the information has changed since we started digging into this a couple of hours ago

Like the OP, I do not know what changed to get the agent into this state.  We've taken this host out of rotation, but the agent is still running on it in this wedged state.

If there's anything we can do to help dig into the issue, please let us know!

**Comment by remh** [comment]
Hi @mccutchen thanks, you are right. We identified the issue and worked on a fix: https://github.com/DataDog/dd-agent/pull/3287

**Comment by mccutchen** [comment]
Awesome, thanks @remh!

**Comment by masci** [comment]
Fixed in 5.12.2
